# Ejercicio 8
### Entrene un combinador lineal que reciba tres dígitos binarios y devuelva el número decimal correspondiente.

a. Utilizando el bias o peso W0

b. Utilizando únicamente las tres entradas correspondientes a los dígitos binarios sin el bias.

### Compare la cantidad de iteraciones necesarias para obtener el vector de pesos correcto en ambos casos. Observe el valor del arco correspondiente al bias en a)

In [4]:
import numpy as np

# Datos de entrada (tres dígitos binarios) y sus correspondientes valores decimales
X = np.array([[0, 0, 0],
              [0, 0, 1],
              [0, 1, 0],
              [0, 1, 1],
              [1, 0, 0],
              [1, 0, 1],
              [1, 1, 0],
              [1, 1, 1]])

# Salidas correspondientes (valores decimales)
Y = np.array([0, 1, 2, 3, 4, 5, 6, 7])

# Agregar una columna de unos para el bias
X_bias = np.c_[np.ones(X.shape[0]), X]

# Inicializar pesos aleatorios
np.random.seed(0)
weights_with_bias = np.random.randn(X_bias.shape[1])

# Inicializar pesos sin bias
weights_no_bias = np.random.randn(X.shape[1])

# Hiperparámetros
learning_rate = 0.01
epochs = 500

# Entrenamiento con bias
for _ in range(epochs):
    prediction_with_bias = np.dot(X_bias, weights_with_bias)
    error_with_bias = prediction_with_bias - Y
    gradient_with_bias = np.dot(X_bias.T, error_with_bias)
    weights_with_bias -= learning_rate * gradient_with_bias

# Entrenamiento sin bias
for _ in range(epochs):
    prediction_no_bias = np.dot(X, weights_no_bias)
    error_no_bias = prediction_no_bias - Y
    gradient_no_bias = np.dot(X.T, error_no_bias)
    weights_no_bias -= learning_rate * gradient_no_bias
print("Pesos con bias:", weights_with_bias)
print("Pesos sin bias:", weights_no_bias)


Pesos con bias: [0.00849301 3.9949951  1.99510088 0.99519369]
Pesos sin bias: [3.99998307 1.99994842 1.00006851]


Los valores de los pesos que se obtuvieron después del entrenamiento son los siguientes:

a) Utilizando el bias (W0):
   - W0 (Bias): 0.00849301
   - W1: 3.9949951
   - W2: 1.99510088
   - W3: 0.99519369

b) Sin utilizar el bias:
   - W1: 3.99998307
   - W2: 1.99994842
   - W3: 1.00006851

Estos pesos representan los parámetros del modelo entrenado que permite convertir tres dígitos binarios en su equivalente decimal a través de una combinación lineal de las entradas. En ambos casos, el modelo es capaz de realizar la conversión correctamente.

La diferencia principal entre los dos conjuntos de pesos es la presencia o ausencia del bias (W0). Cuando se utiliza el bias, el modelo puede representar desplazamientos (offsets) en los datos, lo que puede ser útil en algunos casos. En contraste, el modelo sin bias no tiene en cuenta desplazamientos y solo realiza una combinación lineal de las entradas. La elección de si incluir o no el bias depende de la naturaleza del problema y de los datos que estés tratando de modelar.